### PR TITLE
SAML and OAuth2/OIDC AC with relative callback URLs

### PR DESCRIPTION
**Pull request by johakoch**

### CODE REVIEW

File: eval/context.go [hunk 1]
```diff
@@ -219,31 +221,34 @@ func (c *Context) WithOAuth2(o []*config.OAuth2AC) *Context {
 	return c
 }
 
-// WithSAML initially setup the lib.FnSamlSsoUrl function.
+// WithSAML initially setup the saml configuration.
```
Comment (johakoch) [inline eval/context.go:224]: With only initializing `c.saml` left, is this still a good place? Same with `WithOAuth2()`

File: eval/lib/saml.go [hunk 2]
```diff
@@ -88,6 +94,18 @@ func NewSamlSsoUrlFunction(configs []*config.SAML) function.Function {
 	})
 }
 
+func MakeUrlAbsolute(urlRef string, origin *url.URL) (string, error) {
```
Comment (johakoch) [inline eval/lib/saml.go:97]: Where to put this (generic) function?

**Comment by johakoch** [comment]
We want to be able to use relative URL references as callback URIs for SAML SSO and OAuth2 Authorization Code Flow:

```hcl
  saml "SSO" {
    sp_acs_url = "/saml/acs"
    ...
  }
```
and
```hcl
  beta_oauth2 "google" {
    redirect_uri = "/oidc/cb"
    ...
  }
```
The relative URL references are resolved against the current request URL origin (use `accept_forwarded_url` setting or equivalent run option or env var to change the origin according to `X-Fowarded-*` request headers).

---

**Note:** This only works if the URLs of both the endpoint where the URL creating function (`saml_sso_url()` / `beta_oauth_authorization_url()`) is used and of the callback endpoint share the same origin. So the following will not work because the ports are not the same:
```hcl
server "a" {
  hosts = [ "*:8080" ]
  endpoint "/oidc/start" {
    ... "...beta_oauth_authorization_url("oidc")..." # redirect_uri is http://localhost:8080/oidc/cb
    ...
  }
}
server "b" {
  hosts = [ "*:9000" ]
  endpoint "/oidc/cb" {
    access_control = [ "oidc" ] # redirect_uri is http://localhost:9000/oidc/cb
    ...
  }
}
definitions {
  beta_oauth2 "oidc" {
    redirect_uri = "/oidc/cb"
    ...
  }
}
```

File: eval/lib/saml.go (hunk 2 above)
Comment (afflerbach) [reply eval/lib/saml.go:97]: `./eval/lib/url.go` maybe?

**Review by malud — APPROVED** [review]

**Pull request merged**